File: apps/android/app/src/main/java/com/exponential/app/data/api/PushTokensApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

@Serializable
data class RegisterPushTokenInput(val token: String, val platform: String)

@Serializable
data class UnregisterPushTokenInput(val token: String)

@Serializable
data class PushTokenAck(val ok: Boolean)

@Singleton
class PushTokensApi @Inject constructor(private val trpc: TrpcClient) {
    suspend fun register(token: String) {
        trpc.mutation(
            path = "pushTokens.register",
            input = RegisterPushTokenInput(token = token, platform = "android"),
            inputSerializer = RegisterPushTokenInput.serializer(),
            outputSerializer = PushTokenAck.serializer(),
        )
    }

    suspend fun unregister(token: String) {
        trpc.mutation(
            path = "pushTokens.unregister",
            input = UnregisterPushTokenInput(token),
            inputSerializer = UnregisterPushTokenInput.serializer(),
            outputSerializer = PushTokenAck.serializer(),
        )
    }
}

File: apps/ios/Exponential/Data/API/PushTokensApi.swift
import Foundation

struct RegisterTokenInput: Encodable {
    let token: String
    let platform: String
}

struct UnregisterTokenInput: Encodable {
    let token: String
}

final class PushTokensApi: Sendable {
    private let trpc: TrpcClient

    init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    func register(token: String) async throws {
        try await trpc.mutationVoid(path: "pushTokens.register", input: RegisterTokenInput(token: token, platform: "ios"))
    }

    func unregister(token: String) async throws {
        try await trpc.mutationVoid(path: "pushTokens.unregister", input: UnregisterTokenInput(token: token))
    }
}
